### PR TITLE
Feat/persistant logging system

### PR DESCRIPTION
**Pull request by Basem-Gaber**

### CODE REVIEW

File: opencopilot/utils/logger.py [hunk 1]
```diff
@@ -21,52 +24,94 @@
...
+    rotating_handler.setFormatter(formatter)
+
+    # Create a logger and add the rotating file handler
+    logger = logging.getLogger(__name__)
+    logger.addHandler(rotating_handler)
+    logger.setLevel(log_level)
+
+    return logger
+
+__internal_logger = setup_logger()
```
Comment (oabdelkarim) [inline opencopilot/utils/logger.py:57]: Will this be called in every import?

File: opencopilot/utils/logger.py [hunk 2]
```diff
@@ -21,52 +24,94 @@
 all_colors = [COLOR_BLACK, COLOR_RED, COLOR_GREEN, COLOR_YELLOW, COLOR_BLUE, COLOR_MAGENTA, COLOR_CYAN, COLOR_WHITE, COLOR_LIGHT_GREY, COLOR_DARK_GREY, COLOR_LIGHT_RED, COLOR_LIGHT_GREEN,
               COLOR_LIGHT_YELLOW, COLOR_LIGHT_BLUE, COLOR_LIGHT_MAGENTA, COLOR_LIGHT_CYAN]
 
+def setup_logger():
+    # Read the COPILOT_LOG_LEVEL environment variable
+    log_level_str = os.environ.get("COPILOT_LOG_LEVEL", "INFO")
```
Comment (oabdelkarim) [inline opencopilot/utils/logger.py:29]: Let the default be the least one, not the most verbose one. That way, if not set by mistake, the logs won't explode

File: opencopilot/utils/logger.py [hunk 3]
```diff
@@ -21,52 +24,94 @@
...
+
+    # Convert log level string to integer value
+    log_level = getattr(logging, log_level_str.upper(), logging.INFO)
+
+    # Set up the logging configuration with a rotating file handler
+    log_file = 'logs/app.log'
```
Comment (oabdelkarim) [inline opencopilot/utils/logger.py:35]: We need to make this variable. Not only the file name but the whole mechanims. That way, through the configs, one can set the logger like:

`logging.basicConfig(stream=sys.stdout, level=logging.INFO)`

We can also make it so along with the rotating handler in the configs. The configs are python, so we can intialize the handler there to give freedom for other users of the system to customize it through the configs file or even use stdout/error as above

**Review by oabdelkarim — CHANGES_REQUESTED** [review]

File: opencopilot/utils/logger.py (hunk 1 above)
Comment (oabdelkarim) [reply opencopilot/utils/logger.py:57]: I am not sure whether "add_handler" will keep adding handlers on every call or not! Let us be in the safe side and make sure not to call it if internal_logger is already set

Comment (Basem-Gaber) [reply opencopilot/utils/logger.py]: Done

File: opencopilot/utils/logger.py (hunk 3 above)
Comment (Basem-Gaber) [reply opencopilot/utils/logger.py:38]: Discussed and decided to not do

**Pull request merged**